### PR TITLE
feat: implement new authenticators to handle sdk authentication

### PR DESCRIPTION
**Pull request by dpopp07**
Node changes for new authenticators. Usage examples are in the design wiki.

Breaking changes:
- Base service class requires an Authenticator
- Token managers no longer support access tokens (replaced by Bearer Token Authenticator)
- `Icp4dTokenManagerV1` renamed to `Cp4dTokenManager`
- `IamTokenManagerV1` renamed to `IamTokenManager`
- `JwtTokenManagerV1` renamed to `JwtTokenManager`

Happy to address any feedback

### CODE REVIEW

File: auth/token-managers/iam-token-manager-v1.ts [hunk 1]
```diff
@@ -133,24 +137,28 @@ export class IamTokenManagerV1 extends JwtTokenManagerV1 {
     let clientSecret = 'bx';
```
Comment (padamstx) [inline auth/token-managers/iam-token-manager-v1.ts]: We do not want an Authorization header to be sent to the IAM token server in the event that client id/secret are not specified by the user.

File: auth/authenticators/basic-authenticator.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  password?: string;
+}
+
+export class BasicAuthenticator extends Authenticator implements AuthenticatorInterface {
+  protected requiredOptions = ['username', 'password'];
+  private username: string;
+  private password: string;
+
+  /**
+   * Basic Auth Config Class
```
[the remark this answers is not in the document]
Comment (dpopp07) [reply auth/authenticators/basic-authenticator.ts]: Hhaha rats, an artifact from the work I did on this back before the re-design discussion. I will fix

File: auth/authenticators/cloud-pak-for-data-authenticator.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  url: string;
+}
+
+export class CloudPakForDataAuthenticator extends TokenRequestBasedAuthenticator {
+  protected requiredOptions = ['username', 'password', 'url'];
+  private username: string;
+  private password: string;
+
+  /**
+   * IAM Authenticator Class
```
[the remark this answers is not in the document]
Comment (dpopp07) [reply auth/authenticators/cloud-pak-for-data-authenticator.ts]: You bet 😅

File: auth/token-managers/jwt-token-manager-v1.ts [hunk 4]
```diff
@@ -57,11 +59,12 @@ export class JwtTokenManagerV1 {
...
 
-    if (options.accessToken) {
-      this.userAccessToken = options.accessToken;
-    }
+    // request options
+    this.disableSslVerification = Boolean(options.disableSslVerification);
+    this.headers = options.headers || {};
 
-    this.requestWrapperInstance = new RequestWrapper();
+    // any bonus reqeust options, like `proxy`, will be passed here
```
[the remark this answers is not in the document]
Comment (dpopp07) [reply auth/token-managers/jwt-token-manager-v1.ts]: We expose the config options of the internal request library to the user to give them extended control over their requests. This is how they would specify the proxy, etc. Currently, we use `axios` ([request config options here](https://github.com/axios/axios#request-config)) but we may switch so I wanted to refer to them in a generic way. The word "bonus" was probably not the best choice haha. I can change that

File: lib/base_service.ts [hunk 5]
```diff
@@ -161,331 +58,86 @@ export class BaseService {
...
-    // but prefer credentials provided programatically
-    _options = extend(
-      {},
-      this.getCredentialsFromCloud(this.name),
-      this.getCredentialsFromEnvironment(process.env, this.name),
-      this.getCredentialsFromEnvironment(readCredentialsFile(), this.name),
-      options,
-      _options
-    );
+  private readOptionsFromCredentialsFile() {
```
[the remark this answers is not in the document]
Comment (dpopp07) [reply lib/base_service.ts]: Yes, that was a copy/paste error. Good catch

File: auth/token-managers/iam-token-manager-v1.ts (hunk 1 above)
Comment (dpopp07) [reply auth/token-managers/iam-token-manager-v1.ts]: Oh, so no defaults anymore? Got it. I don't remember that decision. I'll make that change and update the design document just to be clear on it

**Comment by dpopp07** [comment]
Feedback addressed

**Pull request merged**